### PR TITLE
Fix rule dao unindexed select

### PR DESCRIPTION
**Pull request by lchanmann**
## Ref
- https://nanopay.atlassian.net/browse/NP-1688

## Changes
- Add daoKey and enabled indexes to Rule
- Fix scan plan cost function
- Fix simplifying True/False mlangs

### CODE REVIEW

File: src/foam/dao/index/ScanPlan.java [hunk 1]
```diff
@@ -56,7 +56,7 @@ public long calculateCost(PropertyInfo propertyInfo) {
         order_   = null;
       } else {
         sortRequired = true;
-        cost *= Math.log(cost) / Math.log(2);
+        cost = (long) (Math.log(cost) / Math.log(2));
```
Comment (kgrgreer) [inline src/foam/dao/index/ScanPlan.java]: Why this change? Isn't the sorting cost n * log2(n)?

File: src/foam/mlang/mlang.js [hunk 2]
```diff
@@ -508,6 +508,11 @@ foam.CLASS({
       swiftCode: 'return true',
       javaCode: 'return true;'
     },
+    {
+      name: 'partialEval',
+      code: function() { return this },
+      javaCode: 'return foam.mlang.MLang.TRUE;'
```
Comment (kgrgreer) [inline src/foam/mlang/mlang.js:514]: Doesn't this already 'return this', which is the same?

File: src/foam/mlang/mlang.js [hunk 3]
```diff
@@ -533,6 +538,11 @@ foam.CLASS({
       type: 'String',
       javaCode: 'return " 1 <> 1 ";',
       code: function() { return "1 <> 1"; }
+    },
+    {
+      name: 'partialEval',
+      code: function() { return this },
+      javaCode: 'return foam.mlang.MLang.FALSE;'
```
Comment (kgrgreer) [inline src/foam/mlang/mlang.js:545]: Isn't this what already happens?

File: src/foam/dao/index/ScanPlan.java (hunk 1 above)
Comment (lchanmann) [reply src/foam/dao/index/ScanPlan.java]: Is it a search instead of sort? When searching the indexed plan this cost is being used.

File: src/foam/mlang/mlang.js (hunk 2 above)
Comment (lchanmann) [reply src/foam/mlang/mlang.js:514]: No, the singleton axiom doesn't work on java.

    foam.mlang.MLang.TRUE == new True();  // false

File: src/foam/mlang/mlang.js (hunk 3 above)
Comment (lchanmann) [reply src/foam/mlang/mlang.js:545]: No, the singleton axiom doesn't work on java.

File: src/foam/mlang/mlang.js (hunk 2 above)
Comment (kgrgreer) [reply src/foam/mlang/mlang.js:514]: I see. Then that's the real bug.  We need java Singleton support.

File: src/foam/dao/index/ScanPlan.java (hunk 1 above)
Comment (kgrgreer) [reply src/foam/dao/index/ScanPlan.java]: But we said sortRequired = true;
So we will need to sort the values first.

File: src/foam/mlang/mlang.js (hunk 2 above)
Comment (kgrgreer) [reply src/foam/mlang/mlang.js:514]: We can fix this in the future.

File: src/foam/mlang/mlang.js (hunk 3 above)
Comment (kgrgreer) [reply src/foam/mlang/mlang.js:545]: We can fix this in the future.

File: src/foam/dao/index/ScanPlan.java (hunk 1 above)
Comment (lchanmann) [reply src/foam/dao/index/ScanPlan.java]: Fixed in https://github.com/foam-framework/foam2/pull/3791/commits/00294323a4d50cc6a268b2564b4edcd352ddc90e.
